ngboost/distns/multivariate_normal.py

```diff
@@ -1,252 +1,302 @@
+
+            # The remainder is just for utility.
+            self.cov_inv = self.L @ self.L.transpose(0, 2, 1)
+
+            # _cov_mat and _Lcov are place holders, relatively expensive to compute
+            # The inverse of self.cov_inv. Access through self.cov
+            self._cov_mat = None
+            # cholesky factor of _cov_mat, useful for random number generation
+            self._Lcov = None
+            # Saving the pdf constant and means in an accessible format.
+            self.pdf_constant = -self.p / 2 * np.log(2 * np.pi)
+
+        def summaries(self, Y):
+            """
+            Parameters:
+                Y: The data being fit to
+
+            Returns:
+                diff: N x p x1 the residual between the mean and the data
+                eta: N x p which is L@diff
+            """
+            diff = np.expand_dims(self.loc - Y, 2)
+            eta = np.squeeze(np.matmul(self.L.transpose(0, 2, 1), diff))
+            return diff, eta
+
+        def logpdf(self, Y):
+            _, eta = self.summaries(Y)
+            # the exponentiated part of the pdf:
+            p1 = -0.5 * np.sum(np.square(eta), axis=1)
+            p1 = p1.squeeze()
+            # this is the sqrt(determinant(Sigma) component of the pdf
+            p2 = np.sum(np.log(np.diagonal(self.L, axis1=1, axis2=2)), axis=1)
+
+            ret = p1 + p2 + self.pdf_constant
+            return ret
+
+        def fit(Y):
             N, p = Y.shape
             m = Y.mean(axis=0)  # pylint: disable=unexpected-keyword-arg
             diff = Y - m
             sigma = 1 / N * (diff[:, :, None] @ diff[:, None, :]).sum(0)
-            L = sp.linalg.cholesky(sigma, lower=True)
+            L = sp.linalg.cholesky(np.linalg.inv(sigma), lower=True)
+            diag_idx = np.diag_indices(p)
+            L[diag_idx] = np.log(L[diag_idx])
             return np.concatenate([m, L[np.tril_indices(p)]])
+
+        def rv(self):
+            # This is only useful for generating rvs so only compute it in rv.
+            if self._Lcov is None:
+                self._Lcov = np.linalg.cholesky(np.linalg.inv(self.cov_inv))
+
+            u = np.random.normal(loc=0, scale=1, size=(self.N, self.p, 1))
+            sample = np.expand_dims(self.loc, 2) + self._Lcov @ u
+            return np.squeeze(sample)
+
+        def rvs(self, n):
+            return [self.rv() for _ in range(n)]
+
+        def sample(self, n):
+            return self.rvs(n)
+
+        @property
+        def cov(self):
+            # Covariance matrix is for computing the fisher information
+            # If it is singular it is set an extremely large value.
+            # This will probably not affect anything.
+            if self._cov_mat is None:
+                try:
+                    self._cov_mat = np.linalg.inv(self.cov_inv)
+                except np.linalg.LinAlgError:
+                    self._cov_mat = np.identity(self.p) * 1e20
```

Comment:
it's nice that this is handled, but isn't the identity as far from singular as you can get? My intuition would be to set `cov_inv` to something that is as close as possible to its current value but isn't singular and then invert.

Reply:
Good point,

Just ran a couple of examples without this try except and I did not get any errors.

I think when I added it originally it was in __init__ and before adding the small noise to the diagonal of the L matrix in get_chol_factor. 

I'll remove the  except in the next commit